apps/cloudbroker/base/cloudapi__machines/methodclass/cloudapi_machines.py

```diff
@@ -251,47 +260,50 @@ def convertToTemplate(self, machineId, templatename, **kwargs):
         :param basename: snapshot id on which the template is based
         :return True if template was created
         """
-        raise exceptions.NotImplemented("convertToTemplate has been disabled in this release")
+        ctx = kwargs['ctx']
+        user = ctx.env['beaker.session']['user']
+        userobj = j.core.portal.active.auth.getUserInfo(user)
         machine = self._getMachine(machineId)
         origimage = self.models.image.get(machine.imageId)
         if origimage.accountId:
             raise exceptions.Conflict("Can not make template from a machine which was created from a custom template.")
-
-        provider = self.cb.getProvider(machine)
-        node = self.cb.getNode(machine.referenceId, provider)
+        provider, node, _ = self.cb.getProviderAndNode(machineId)
+        firstdisk = self.models.disk.get(machine.disks[0])
         cloudspace = self.models.cloudspace.get(machine.cloudspaceId)
         image = self.models.image.new()
-        image.name = templatename
+        image.name = templateName
         image.type = 'Custom Templates'
         image.username = machine.accounts[0].login
         image.password = machine.accounts[0].password
-        m = {}
-        m['stackId'] = machine.stackId
-        m['disks'] = machine.disks
-        m['sizeId'] = machine.sizeId
-        firstdisk = self.models.disk.get(machine.disks[0])
         image.size = firstdisk.sizeMax
         image.accountId = cloudspace.accountId
         image.status = 'CREATING'
         imageid = self.models.image.set(image)[0]
         image.id = imageid
         image.gid = cloudspace.gid
         try:
-            referenceId = provider.ex_create_template(node, templatename)
+            volume = provider.create_volume(image.size, templateName)
+            referenceId = provider.ex_create_template(node, templateName, volume.vdiskguid)
         except:
```

Comment:
you need to make an eco here which you pass the referenceId in the email so they can contact support also for us they have some trace something went wrong